File: JavaScript/numparser.js
eval (GetFileContent ('lc/ru.js'));
eval (GetFileContent ('helpers.js'));

function NumeralProp (value, level, is_multiplier, is_eleven_to_nineteen) 
{    
    this.value = value;
    this.level = level;
    this.is_multiplier = is_multiplier;
    this.is_eleven_to_nineteen = (is_eleven_to_nineteen === undefined)? false : is_eleven_to_nineteen;
}    

function NumericToken (numeral, error) 
{
    this.numeral = numeral;
    this.error   = (error === undefined)? 0 : error;
    this.is_significant = false;
}

function ParserResult (value, error) 
{
    this.value = value;
    this.error = (error === undefined)? 0 : error;
}

function Parser () {

    var Parent = new Ru ();

    function CleanPunctuation (In)
    {
       var Punctuation = '!"#$%&\'()*+, -./:;<=>?@[\]^_`{|}~';
       var StartPos    = 0;
       var EndPos      = In.length;

       for (;StartPos < EndPos; StartPos++) {
           if (Punctuation.indexOf (In.charAt (StartPos)) === -1) break;
       }

       for (;EndPos != 0; EndPos--) {
          if (Punctuation.indexOf (In.charAt (EndPos)) === -1) break;
       }

       return In.substring (StartPos, EndPos + 1);
    }

    function GetKeys (Arr) 
    {
        var KeysTokens = [];
        for (var Key in Arr) KeysTokens.push (Key);
        return KeysTokens;
    }

    function inArray (Arr, Prop)
    {
        var Count = -1;
        for (var i in Arr) {
            Count++;
            if (Arr[i] == Prop) return Count;
        }
        return -1;
    }

    Parent.get_token_sum_error_from_lists = function (token) 
    {
        var token_sum_error = 0;

        if (token instanceof NumericToken) token_sum_error += token.error
        else {
            for (var token_idx in token) {
                var sub_token = token[token_idx];
                token_sum_error += this.get_token_sum_error_from_lists (sub_token)
            }
        }

        return token_sum_error;
    }

    Parent.parse_tokens = function (text_line, fraction) 
    {
        this.fraction   = (fraction === undefined)? false : fraction;
        var keys_tokens = GetKeys (this.tokens);

        if (inArray (keys_tokens, text_line) != -1) {
            return [new NumericToken (this.tokens[text_line])];
        }
        else if (fraction) return [new NumericToken (this.tokens_fractions[text_line])];
        else return [null];
    }

    Parent.parse  = function (text) 
    {
        var text = Trim (text).toLowerCase ();

        if (text.length == 0) return ParserResult (0, 1);

        // Разбиваем текст на токены
        var raw_token_list        = text.split (/\s+/);
        var all_token_list        = [];
        var token_list            = [];
        var result_text_list      = [];
        var left_space_for_number = false;
        var current_level         = 0;

        // Обрабатываем токены
        for (var token_idx in raw_token_list) {

            token_idx     = +token_idx;
            var raw_token = raw_token_list[token_idx]

            var clean_token        = CleanPunctuation  (raw_token);
            var current_token_list = this.parse_tokens (clean_token, false)

            //WScript.Echo (ShowObject (current_token_list));
            
            // Получаем массив числительных языка
            var keys_tokens    = GetKeys (this.tokens);
            var keys_fractions = GetKeys (this.tokens_fractions);
            
            // Определение дробного числа:
            if (inArray (['целых', 'целой', 'целым', 'целая'], clean_token) != -1 && token_idx != 0) {
                
                if ((inArray (keys_tokens, raw_token_list[token_idx - 1]) != -1  &&
                     inArray (keys_tokens, raw_token_list[token_idx + 1]) != -1) ||
                     raw_token_list[token_idx + 1] == "и") {
                        
                    if (inArray (keys_fractions, raw_token_list[token_idx + 2]) != -1 ||
                        inArray (keys_fractions, raw_token_list[token_idx + 3]) != -1 ||
                        inArray (keys_fractions, raw_token_list[token_idx + 4]) != -1) {

                        current_token_list = this.parse_tokens (clean_token, true);
                    }
                }
            }

            // Обработка первого порядка:
            if (inArray (["десятых", "десятой", "десятым", "десятая"], clean_token) != -1 && token_idx != 0) {

                if (inArray (keys_tokens, raw_token_list[token_idx - 1]) != -1) {
                    current_token_list = this.parse_tokens (clean_token, true)
                }
            }

            // Обработка второго порядка:
            if (inArray (["сотых", "сотой", "сотым", "сотая"], clean_token) != -1 && token_idx != 0) {

                if (inArray (keys_tokens, raw_token_list[token_idx - 1]) != -1) {
                    current_token_list = this.parse_tokens (clean_token, true)
                }
            }

            // Обработка третьего порядка:
            if (inArray (["тысячных", "тысячной", "тысячным", "тысячная"], clean_token) != -1 && token_idx != 0) {

                if (inArray (keys_tokens, raw_token_list[token_idx - 1]) != -1) {
                    current_token_list = this.parse_tokens (clean_token, true)
                }
            }

            // Обработка четвёртого порядка:
            if (inArray (["десятитысячных", "десятитысячной", "десятитысячным", "десятитысячная"], clean_token) != -1 && token_idx != 0) {

                if (inArray (keys_tokens, raw_token_list[token_idx - 1]) != -1) {
                    current_token_list = this.parse_tokens (clean_token, true)
                }
            }

            // Простые числительные - исключения
            if (raw_token == "тысяча") {

                if (token_idx == 0 || raw_token_list[token_idx - 1] != "одна") {

                    current_token_list = [new NumericToken (Numeral (1, 1, false)), current_token_list[0]];
                    current_level      = 1;

                    if (token_list.length > 0) {
                        all_token_list.push (token_list);
                        token_list = [];
                    }

                    left_space_for_number = false;
                }
            }

            if (raw_token == "тысячная") {

                if (token_idx == 0 || raw_token_list[token_idx - 1] != "одна") {

                    current_token_list = [new NumericToken (Numeral (1000, 1, false))];
                    current_level      = 0;

                    if (token_list.length > 0) {
                        all_token_list.push (token_list);
                        token_list = [];
                    }

                    left_space_for_number = false;
                }
            }

            if (raw_token == "сотая") {

                if (token_idx == 0 || raw_token_list[token_idx - 1] != "одна") {

                    current_token_list = [new NumericToken,(Numeral (100, 1, false))];
                    current_level      = 0;

                    if (token_list.length > 0) {
                        all_token_list.push (token_list);
                        token_list = [];
                    }

                    left_space_for_number = false;
                }
            }

            if (raw_token == "десятая") {

                if (token_idx == 0 || raw_token_list[token_idx - 1] != "одна") {

                    current_token_list = [new NumericToken (Numeral (10, 1, false))];
                    current_level      = 0;

                    if (token_list.length > 0) {
                        all_token_list.push (token_list);
                        token_list = [];
                    }

                    left_space_for_number = false;
                }
            }

            if (raw_token == "ноль") {

                if (token_idx != 0 || raw_token_list[token_idx - 1] == "ноль") {

                    current_token_list = [new NumericToken (Numeral (0, 1, false)), current_token_list[0]];
                    current_level      = 0;

                    if (token_list.length > 0) {
                        all_token_list.push (token_list);
                        token_list = [];
                    }

                    left_space_for_number = false;
                }
            }
            
            //WScript.Echo (ShowObject (current_token_list));

            // Обработка целой части числа
            if (current_token_list[0] != null) {

                var list_size  = 0;
                previous_level = current_level;

                for (var current_token_list_idx in current_token_list) {
                    if (current_token_list.hasOwnProperty (current_token_list_idx)) list_size++;
                }

                current_level         = current_token_list[list_size - 1].numeral.level;
                is_eleven_to_nineteen = current_token_list[0].numeral.is_eleven_to_nineteen

                if (current_level != 0 && previous_level != 0 && (
                        (previous_level < current_level && current_level <= 3) || 
                        (current_level == previous_level)                      || 
                        (current_level < previous_level && previous_level <= 2 && is_eleven_to_nineteen))) {
                    
                    all_token_list.push (token_list);
                    token_list = [];
                    left_space_for_number = false;
                    
                }
            }

            //WScript.Echo (ShowObject (all_token_list));

            var bad_tokens = true;

            for (var current_idx in current_token_list) {
                var current_token = current_token_list[current_idx];

                if (current_token == null) break;

                if (current_token.error <= this.max_token_error) {
                    bad_tokens = false;
                    break;
                }
            }

            if (bad_tokens) current_token_list = [];

            if (current_token_list[0] != null) {

                if (!left_space_for_number) {
                    result_text_list.push ('');
                    left_space_for_number = true;
                }

                for (var current_idx in current_token_list) {
                    var current_token = current_token_list[current_idx];
                    token_list.push (current_token);
                }
            }
            else {
                result_text_list.push (raw_token);
                left_space_for_number = false;
                current_level = 0;

                if (token_list.length > 0) {
                    all_token_list.push (token_list);
                    token_list = [];
                }
            }
        }

        if (token_list.length > 0) {
            all_token_list.push (token_list);
            token_list = [];
        }

        var parser_result_list = [];

        // Собираем токены в число
        for (var all_token_list_idx in all_token_list) {

            var global_level   = null;
            var local_level    = null;
            var global_value   = null;
            var local_value    = null;
            var critical_error = false;

            var token_list  = all_token_list[all_token_list_idx];
            var token_count = token_list.length;

            //WScript.Echo (ShowObject (token_list));

            for (var current_idx in token_list) {

                var current_token = token_list[current_idx];
                var current_error = this.get_token_sum_error_from_lists (current_token);

                if (current_error > this.max_token_error) {
                    continue
                }

                //WScript.Echo (ShowObject (current_token));

                var value      = current_token.numeral.value;
                var level      = current_token.numeral.level;
                var multiplier = current_token.numeral.is_multiplier;

                if (multiplier) {

                    if (global_level == null) {
                        //WScript.Echo ('Множитель - первый');

                        if (local_level == null) global_value = global_value + value;
                        else global_value = Math.round (local_value * value * 100000) / 100000;

                        global_level = level;
                        local_value  = null;
                        local_level  = null;

                        current_token.is_significant = true;
                    }
                    else if (global_level > level) {
                        //WScript.Echo ('Множитель - приставной');

                        if (local_level == null) global_value = global_value + value;
                        else{
                            if      (value >= 0.01)   global_value = Math.round ((global_value + local_value * value) * 100) / 100;
                            else if (value == 0.001)  global_value = Math.round ((global_value + local_value * value) * 1000) / 1000;
                            else if (value == 0.0001) global_value = Math.round ((global_value + local_value * value) * 10000) / 10000;
                        }

                        global_level = level;
                        local_value  = null;
                        local_level  = null;

                        current_token.is_significant = true;
                    }
                    else {
                        //WScript.Echo ('Множитель - ошибка');

                        current_token.error = 1;
                        critical_error      = true;

                        current_token.is_significant = true;
                    }
                }
                else {

                    if (local_level == null) {
                        //WScript.Echo ('Цифра - первая');

                        local_value = value;
                        local_level = level;

                        current_token.is_significant = true;
                    }
                    else if (local_level > level) {
                        //WScript.Echo ('Цифра - приставная');

                        local_value = local_value + value;
                        local_level = level;
                        
                        current_token.is_significant = true;
                    }
                    else {
                        //WScript.Echo ('Цифра - Ошибка');

                        current_token.error = 1;
                        critical_error      = true;

                        current_token.is_significant = true;
                    }
                }
            }

            // Считаем общий уровень ошибки
            if (token_count == 0) total_error = 1;
            else {
                total_error             = 0;
                significant_token_count = 0;

                for (var idx in token_list) {

                    current_token = token_list[idx];

                    if (current_token.is_significant) {
                        total_error += current_token.error;
                        significant_token_count += 1;
                    }
                }

                total_error /= significant_token_count;
            }

            // Критическая ошибка
            if (critical_error) {
                if (total_error >= 0.5) total_error  = 1;
                else                    total_error *= 2;
            }

            result_value = 0;

            if (global_value != null) result_value += global_value;
            if (local_value  != null) result_value += local_value;

            parser_result_list.push (new ParserResult (result_value, total_error));
        }

        //WScript.Echo (ShowObject (parser_result_list));
        //WScript.Echo (ShowObject (result_text_list));

        return [parser_result_list, result_text_list];
    }

    Parent.constructor = arguments.callee;
    return Parent;
}

File: Python/numparser.py
import string, re
import numpy as np

from helpers import Helpers as hp
from lc.numeral import Numeral
from lc.ru import Ru

class NumericToken:
    def __init__(self, numeral, error=0):
        self.numeral = numeral
        self.error = error
        self.is_significant = False

    def print(self, indent):
        return '\n' \
               + indent + '{numeral = ' + str(self.numeral.print (indent + '   ')) + ',\n' + ' ' \
               + indent + 'error = '   +  str(self.error) + ',\n' + ' ' \
               + indent + 'is_significant = ' + str(self.is_significant) + '}'

class ParserResult:
    def __init__(self, value, error=0):
        self.value = value
        self.error = error

    def print(self, indent):
        return '{value = ' + str(self.value) + ',\n' + ' ' \
               + indent + 'error = ' +  str(self.error) +  '}'

class Parser:
    def __init__(self):
        self.ru = Ru()
        self.tokens = self.ru.tokens
        self.tokens_fractions = self.ru.tokens_fractions
        self.max_token_error = 0.3

    def get_token_sum_error_from_lists(self, token):
        token_sum_error = 0

        if isinstance(token, NumericToken):
            token_sum_error += token.error

        else:
            for sub_token in token:
                token_sum_error += self.get_token_sum_error_from_lists(sub_token)

        return token_sum_error

    def parse_tokens(self, text_line, level, fraction=False):
        if text_line in self.tokens.keys():
            return [NumericToken(self.tokens[text_line])]
        elif fraction:
            return [NumericToken(self.tokens_fractions[text_line])]
        else:
            return [None]

    def parse(self, text):
        text = text.strip().lower()

        if len(text) == 0:
            return ParserResult(value=0, error=1)

        # Разбиваем текст на токены
        raw_token_list = re.split(r"\s+", text)
        all_token_list = []
        token_list = []
        result_text_list = []
        left_space_for_number = False
        current_level = 0

        # Обрабатываем токены
        for token_idx, raw_token in enumerate(raw_token_list):
            clean_token = raw_token.strip(string.punctuation)
            current_token_list = self.parse_tokens(clean_token, 0)
            #print (hp.show_object(current_token_list))

            # Определение дробного числа:
            if clean_token in ["целых", "целой", "целым", "целая"] and token_idx != 0:

                #print (raw_token_list)

                try:
                    if raw_token_list[token_idx - 1] in self.tokens \
                            and raw_token_list[token_idx + 1] in self.tokens \
                            or raw_token_list[token_idx + 1] == "и":
                        if raw_token_list[token_idx + 2] in self.tokens_fractions \
                                or raw_token_list[token_idx + 3] in self.tokens_fractions \
                                or raw_token_list[token_idx + 4] in self.tokens_fractions:
                            current_token_list = self.parse_tokens(clean_token, 0, fraction=True)

                except IndexError:
                    pass

            # Обработка первого порядка:
            if clean_token in ["десятых", "десятой", "десятым", "десятая"] and token_idx != 0:

                if raw_token_list[token_idx - 1] in self.tokens:
                    current_token_list = self.parse_tokens(clean_token, 0, fraction=True)

            # Обработка второго порядка:
            if clean_token in ["сотых", "сотой", "сотым", "сотая"] and token_idx != 0:

                if raw_token_list[token_idx - 1] in self.tokens:
                    current_token_list = self.parse_tokens(clean_token, 0, fraction=True)

            # Обработка третьего порядка:
            if clean_token in ["тысячных", "тысячной", "тысячным", "тысячная"] and token_idx != 0:

                if raw_token_list[token_idx - 1] in self.tokens:
                    current_token_list = self.parse_tokens(clean_token, 0, fraction=True)

            # Обработка четвёртого порядка:
            if clean_token in ["десятитысячных", "десятитысячной", "десятитысячным", "десятитысячная"] and token_idx != 0:

                if raw_token_list[token_idx - 1] in self.tokens:
                    current_token_list = self.parse_tokens(clean_token, 0, fraction=True)

            if raw_token == "тысяча":
                if token_idx == 0 or raw_token_list[token_idx - 1] != "одна":
                    current_token_list = [NumericToken(numeral=Numeral(1, 1, False)), current_token_list[0]]
                    current_level = 1

                    if len(token_list) > 0:
                        all_token_list.append(token_list)
                        token_list = []

                    left_space_for_number = False

            if raw_token == "тысячная":
                if token_idx == 0 or raw_token_list[token_idx - 1] != "одна":
                    current_token_list = [NumericToken(numeral=Numeral(1000, 1, False))]
                    current_level = 0

                    if len(token_list) > 0:
                        all_token_list.append(token_list)
                        token_list = []

                    left_space_for_number = False

            if raw_token == "десятая":
                if token_idx == 0 or raw_token_list[token_idx - 1] != "одна":
                    current_token_list = [NumericToken(numeral=Numeral(10, 1, False))]
                    current_level = 0

                    if len(token_list) > 0:
                        all_token_list.append(token_list)
                        token_list = []

                    left_space_for_number = False

            if raw_token == "сотая":
                if token_idx == 0 or raw_token_list[token_idx - 1] != "одна":
                    current_token_list = [NumericToken(numeral=Numeral(100, 1, False))]
                    current_level = 0

                    if len(token_list) > 0:
                        all_token_list.append(token_list)
                        token_list = []

                    left_space_for_number = False

            if raw_token == "ноль":
                if token_idx != 0 or raw_token_list[token_idx - 1] == "ноль":
                    current_token_list = [NumericToken(numeral=Numeral(0, 1, False)), current_token_list[0]]
                    current_level = 0

                    if len(token_list) > 0:
                        all_token_list.append(token_list)
                        token_list = []

                    left_space_for_number = False

            if current_token_list[0] is not None:
                #print(hp.show_object(current_token_list))
                previous_level = current_level
                current_level = current_token_list[len(current_token_list) - 1].numeral.level
                is_eleven_to_nineteen = current_token_list[0].numeral.is_eleven_to_nineteen

                #print(current_level)
                #print(previous_level)

                if current_level != 0 and previous_level != 0 and (
                        (previous_level < current_level <= 3) or 
                        (current_level == previous_level)     or 
                        (current_level < previous_level <= 2 and is_eleven_to_nineteen)):

                    all_token_list.append(token_list)
                    token_list = []
                    left_space_for_number = False

            #print(hp.show_object(all_token_list))

            bad_tokens = True

            for current_token in current_token_list:
                if current_token is None:
                    break

                if current_token.error <= self.max_token_error:
                    bad_tokens = False
                    break

            #print(bad_tokens)

            if bad_tokens is True:
                del current_token_list
                current_token_list = [None]

            if current_token_list[0] is not None:
                if left_space_for_number is False:
                    result_text_list.append("")
                    left_space_for_number = True

                for current_token in current_token_list:
                    token_list.append(current_token)
                    #print(hp.show_object(token_list))


            else:
                result_text_list.append(raw_token)
                left_space_for_number = False
                current_level = 0

                if len(token_list) > 0:
                    all_token_list.append(token_list)
                    token_list = []

        if len(token_list) > 0:
            all_token_list.append(token_list)
            token_list = []

        #pprint(vars(all_token_list[0][1].numeral))
        parser_result_list = []

        for token_list in all_token_list:
            global_level = None
            local_level = None
            global_value = None
            local_value = None
            critical_error = False

            token_count = len(token_list)

            for current_token in token_list:
                current_error = self.get_token_sum_error_from_lists(current_token)

                if current_error > self.max_token_error:
                    continue

                value = current_token.numeral.value
                level = current_token.numeral.level
                multiplier = current_token.numeral.is_multiplier

                if multiplier:
                    if global_level is None:
                        if local_level is None:
                            global_value = value
                        else:
                            global_value = np.round(local_value * value, 5)

                        global_level = level
                        local_value = None
                        local_level = None

                        current_token.is_significant = True

                    elif global_level > level:
                        if local_level is None:
                            global_value = global_value + value
                        else:
                            if value >= 0.1:
                                global_value = np.round((global_value + local_value * value), 1)
                            elif value == 0.01:
                                global_value = np.round((global_value + local_value * value), 2)
                            elif value == 0.001:
                                global_value = np.round((global_value + local_value * value), 3)
                            elif value == 0.0001:
                                global_value = np.round((global_value + local_value * value), 4)

                        global_level = level
                        local_value = None
                        local_level = None

                        current_token.is_significant = True

                    else:
                        # Ошибка несоответствия уровней
                        current_token.error = 1
                        current_token.is_significant = True
                        critical_error = True

                else:
                    # Простое числительное
                    if local_level is None:
                        local_value = value
                        local_level = level

                        current_token.is_significant = True

                    elif local_level > level:
                        local_value = local_value + value
                        local_level = level

                        current_token.is_significant = True

                    else:
                        # Ошибка несоответствия уровней
                        current_token.error = 1
                        current_token.is_significant = True
                        critical_error = True

            # Считаем общий уровень ошибки
            if token_count == 0:
                total_error = 1

            else:
                total_error = 0
                significant_token_count = 0

                for current_token in token_list:
                    if current_token.is_significant:
                        total_error += current_token.error
                        significant_token_count += 1

                total_error /= significant_token_count

            if critical_error:
                # Имела место критическая ошибка
                if total_error >= 0.5:
                    total_error = 1

                else:
                    total_error *= 2

            result_value = 0

            if global_value is not None:
                result_value += global_value

            if local_value is not None:
                result_value += local_value

            parser_result_list.append(ParserResult(result_value, total_error))

            #print (hp.show_object(parser_result_list))
            #print (hp.show_object(result_text_list))

        return parser_result_list, result_text_list

    def convert(self, text_line):
        if not text_line:
            return text_line

        parsed_list, result_text_list = self.parse(text=text_line)
        converted_text = ""
        parsed_idx = 0
        result_text_list_len = len(result_text_list)

        for i, element in enumerate(result_text_list):
            if element == "":
                converted_text += str(parsed_list[parsed_idx].value)
                parsed_idx += 1

            else:
                converted_text += element

            if i < result_text_list_len - 1:
                converted_text += " "

        return converted_text

if __name__ == "__main__":
    parser = Parser()

    while True:
        text_line = input("Введите ваш текст:\n")
        converted_line = parser.convert(text_line)
        print(f"\nРаспознанное: {converted_line}\n\n")